### PR TITLE
ICS28: Enable the removal of a consumer chain from the provider

### PR DESCRIPTION
**Pull request by mpoke**
Closes https://github.com/cosmos/ibc/issues/651

Closes https://github.com/cosmos/ibc/issues/669
- [x] handle timeouts on the consumer chain side (see TODOs [here](https://github.com/cosmos/ibc/blob/marius/651-ccv-remove-consumer/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-ccf-tomat1) and [here](https://github.com/cosmos/ibc/blob/marius/651-ccv-remove-consumer/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-ccf-toslash1))

### CODE REVIEW

File: spec/app/ics-028-cross-chain-validation/technical_specification.md [hunk 1]
```diff
@@ -161,19 +161,29 @@ The CCV module is initialized through the `InitGenesis` method when the chain is
...
   ```
   - `chainId` is the proposed chain ID of the new consumer chain. It must be different from all other consumer chain IDs of the executing provider chain.
   - `initialHeight` is the proposed initial height of new consumer chain. Note that `Height` is defined in [ICS 7](../../client/ics-007-tendermint-client). For a completely new chain, `initialHeight = {0,1}`; however, it may be different if the chain converts to a consumer chain.
-  - `spawnTime` is the time on the provider chain at which the consumer chain genesis is finalized and all validators will be responsible for starting their consumer chain validator node.
+  - `spawnTime` is the time on the provider chain at which the consumer chain genesis is finalized and all validators are responsible to start their consumer chain validator node.
+  - `lockUnbondingOnTimeout` is a boolean value that indicates whether the funds corresponding to the outstanding unbonding operations are to be released in case of a timeout. In case `lockUnbondingOnTimeout == true`, a governance proposal to stop the timed out consumer chain would be necessary to release the locked funds. 
+  ```typescript
+  interface StopConsumerChainProposal {
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/technical_specification.md:180]: What do you mean by enforcement? If the proposal passes, then the consumer chain is removed once `stopTime` elapses. Validators can still run full nodes of the consumer chain without being penalized. So in that sense, it is indeed a signaling proposal. The only thing enforced is that validators are no longer slashed for not validating on the consumer chain. Should I make this clearer in the spec?

**Comment by mpoke** [comment]
@AdityaSripal Regarding handling timeouts on the consumer chain side (see TODOs [here](https://github.com/cosmos/ibc/blob/marius/651-ccv-remove-consumer/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-ccf-tomat1) and [here](https://github.com/cosmos/ibc/blob/marius/651-ccv-remove-consumer/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-ccf-toslash1)), is there anything else that must be done besides shutting down the chain? 

How is the provider chain notified that the consumer chain shut down? I guess that the provider either receives a `MsgTimeoutOnClose` message from the relayer proving that the consumer closed the CCV channel or waits until its own packets timeout, right? 

Finally, how can the consumer chain move its CCV channel to CLOSED and shut down at the same time?

File: spec/app/ics-028-cross-chain-validation/technical_specification.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdityaSripal) [reply spec/app/ics-028-cross-chain-validation/technical_specification.md:180]: No one gets slashed. The child chain could effectively continue running as an independent chai

**Comment by AdityaSripal** [comment]
So long as our side does `ChanCloseInit`, a relayer should be able to do TimeoutOnClose on the counterpary

**Comment by mpoke** [comment]
> So long as our side does `ChanCloseInit`, a relayer should be able to do TimeoutOnClose on the counterpary

If the consumer chain executes `onTimeoutPacket()`, the IBC logic will move the channel to `CLOSED` (e.g., [here](https://github.com/cosmos/ibc-go/blob/86638c29107d8a0ad0c70ad1676523edfeea254d/modules/core/04-channel/keeper/timeout.go#L157)). This should be sufficient for the relayer to do TimeoutOnClose on the counterpary. However, when is the consumer chain shutting down? It cannot be in `onTimeoutPacket()`, since the channel will not chage to `CLOSED`. We could add the following lines to the `BeginBlock()` of the consumer CCV module 
```typescript
channel = provableStore.get(channelPath(ConsumerPortId, providerChannel))
abortSystemUnless(providerChannel == "" OR channel.state != CLOSED)
```
What do you say?

**Comment by AdityaSripal** [comment]
I don't know if there's a way to shut down the consumer chain within the state machine itself. I suppose validators can simply shut down their nodes once the channel has closed (block gets committed). Though they still need to make their nodes available so relayers can query proofs of channel close

**Comment by mpoke** [comment]
> I don't know if there's a way to shut down the consumer chain within the state machine itself. I suppose validators can simply shut down their nodes once the channel has closed (block gets committed). Though they still need to make their nodes available so relayers can query proofs of channel close

This is important since as soon as the CCV channel is closed, the consumer chain is no longer secured, i.e., the validators can do whatever they want. We should be able to cleanly shut down (e.g., export the state and no longer accept TXs).

**Comment by mpoke** [comment]
I tried to deal with shutting down the consumer chain in the commit https://github.com/cosmos/ibc/pull/707/commits/ea6919b2eb8ec66c97f9597eae1c4089b33c462a. @AdityaSripal Could you please have a quick look?

**Review by AdityaSripal — APPROVED** [review]
Nice work Marius!

File: spec/app/ics-028-cross-chain-validation/technical_specification.md [hunk 2]
```diff
@@ -161,19 +161,39 @@ The CCV module is initialized through the `InitGenesis` method when the chain is
...
 While the structure of governance proposals is specific to every ABCI application (for an example, see the `Proposal` interface in the [Governance module documentation](https://docs.cosmos.network/v0.44/modules/gov/) of Cosmos SDK),
-this specification expects the following fields to be part of every proposal to spawn a new consumer chain:
+this specification expects the following fields to be part of the proposals to spawn new consumer chains (i.e., `SpawnConsumerChainProposal`) and to stop existing ones (i.e., `StopConsumerChainProposal`):
   ```typescript
-  interface CreateConsumerChainProposal {
+  interface SpawnConsumerChainProposal {
     chainId: string
     initialHeight: Height
     spawnTime: Timestamp
+    lockUnbondingOnTimeout: Bool
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/technical_specification.md:172]: Good point. I'll look into it.

**Comment by mpoke** [comment]
@josef-widder I added a note discussing the implications of calling `StopConsumerChain` with `lockUnbonding` set to false (see https://github.com/cosmos/ibc/pull/707/commits/3f24b4bde18dc6ecd2080b22efb25d253700fae6).

**Pull request merged**